workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ComposeLifecycleOwner.kt

```diff
@@ -0,0 +1,75 @@
+package com.squareup.workflow1.ui.compose
+
+import androidx.compose.runtime.RememberObserver
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleEventObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.LifecycleRegistry
+
+/**
+ * A custom [LifecycleOwner] that synchronizes its lifecycle with a parent [Lifecycle] and
+ * integrates with Jetpack Compose's lifecycle through [RememberObserver].
+ *
+ * **Purpose:**
+ *
+ * - Ensures that any lifecycle-aware components within a composable function have a lifecycle that
+ *   accurately reflects both the parent lifecycle and the composable's own lifecycle.
+ * - Manages lifecycle transitions and observer registration/removal to prevent memory leaks and
+ *   ensure proper cleanup when the composable leaves the composition.
+ *
+ * **Key Features:**
+ *
+ * - **Lifecycle Synchronization:** Mirrors lifecycle events from the provided `parentLifecycle` to
+ *   its own [LifecycleRegistry], ensuring consistent state transitions.
+ * - **Compose Integration:** Implements [RememberObserver] to align with the composable's lifecycle
+ *   in the Compose memory model.
+ * - **Automatic Observer Management:** Adds and removes a [LifecycleEventObserver] to the parent
+ *   lifecycle, preventing leaks and ensuring proper disposal.
+ * - **State Transition Safety:** Carefully manages lifecycle state changes to avoid illegal
+ *   transitions, especially during destruction.
+ *
+ * **Usage Notes:**
+ *
+ * - Should be used in conjunction with `remember` and provided the `parentLifecycle` as a key to
+ *   ensure it updates correctly when the parent lifecycle changes.
+ * - By integrating with Compose's lifecycle, it ensures that resources are properly released when
+ *   the composable leaves the composition.
+ *
+ * @param parentLifecycle The parent [Lifecycle] with which this lifecycle owner should synchronize.
+ */
+internal class ComposeLifecycleOwner(
+  private val parentLifecycle: Lifecycle
+) : LifecycleOwner, RememberObserver {
+
+  private val registry = LifecycleRegistry(this)
+  override val lifecycle: Lifecycle
+    get() = registry
+
+  private val parentObserver = LifecycleEventObserver { _, event ->
+    // Any time the parent lifecycle changes state, perform the same change on our lifecycle.
+    registry.handleLifecycleEvent(event)
+  }
+
+  init {
+    // We do this right away to ensure that the parent lifecycle is observed as soon as possible
+    // to prevent any inconsistencies.
+    parentLifecycle.addObserver(parentObserver)
+  }
+
+  override fun onRemembered() {
+
+  }
+
+  override fun onAbandoned() {
+    onForgotten()
+  }
+
+  override fun onForgotten() {
+    parentLifecycle.removeObserver(parentObserver)
+
+    // If we're leaving the composition, ensure the lifecycle is cleaned up
+    if (registry.currentState != Lifecycle.State.INITIALIZED) {
+      registry.currentState = Lifecycle.State.DESTROYED
+    }
```

Comment:
One behavior change this introduces that we might not want is that it will destroy the child lifecycle before sending a new one, if the parent lifecycle instance changes. I don't think the parent instance should ever change though, unless `movableContentOf` is used or something, so it's probably not gonna be an issue immediately.

Reply:
Interesting, i guess that means we can just not key off parent lifecycle on the remember call

Reply:
I made the update to ensure this behavior didn't change since i'd hope it would work with `movableContentOf` & wrote a test for it too: https://github.com/square/workflow-kotlin/pull/1234/files#diff-be84e80fcb3d58b96b21bf4311c0bcdc15c6f2d0d90b6628ce1f220bd012301eR85

Reply:
I'm guessing you didn't push that yet? It doesn't look changed to me. This should push this code to be even more similar to the View one.

Reply:
yeah i removed `parentLifecycle` from the remember key. check out: https://github.com/square/workflow-kotlin/pull/1234/files#diff-be84e80fcb3d58b96b21bf4311c0bcdc15c6f2d0d90b6628ce1f220bd012301eR76

it will retain the same  instance of `LifecycleOwner` regardless of parent lifecycle changing.

Reply:
Oh, that's also incorrect because if the parent changes then we'll be syncing to the wrong one. You need to add some code so the remember function can update the `parentLifecycle` on this new class, and make sure we remove/add observers correctly when that happens (without necessarily destroying the child).

Reply:
i think i got what you mean now: https://github.com/square/workflow-kotlin/pull/1234/commits/953e2ab89b67999574ba5953255318a4d3bf8636